### PR TITLE
Fix 10 alerts from LGTM (#171)

### LINKED ISSUE #171: Mitigating security vulnerability analysis from LGTM
_Originally related to PR #170 proposed by @tdrozdovsky_

### PR DESCRIPTION
**Pull request by tdrozdovsky**
Signed-off-by: Taras Drozdovskyi <t.drozdovsky@samsung.com>

# Description

Mitigation of alerts from the LGTM. The local variable `err` in the `switch`operator does not pass the error value to the result of the function, therefore, when an error occurs, the function does not return an error.

Fixes #171

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

```
Example
  - [x] Unittest
 ```

**Test Configuration**:
* Firmware version: Ubuntu 16.04
* Hardware: x86-64
* Toolchain: Docker and Go recommended versions
* Edge Orchestration Release: Coconut

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by MoonkiHong — COMMENTED** [review]
@tdrozdovsky Thank you for this PR. The suggested one is to resolve the overwritten attribute `err` in the given functions, am I correct? (`err` vs. `errn`) We are not sure how LGTM works in their analysis, so could you give us a little further information like : LGTM analysis result in your forked branch?

**Comment by MoonkiHong** [comment]
> @tdrozdovsky Thank you for this PR. The suggested one is to resolve the overwritten attribute `err` in the given functions, am I correct? (`err` vs. `errn`) We are not sure how LGTM works in their analysis, so could you give us a little further information like : LGTM analysis result in your forked branch?

@tdrozdovsky Sorry that making you confused. LGTM does not provide any analysis on the forked version. So my suggestion is:

1. Review and revise this PR here;
2. (If we adopt this PR) let us see the updated LGTM analysis result;
3-1. (If it works) Good;
3-2. (If it does not work) We might revoke this PR and find other resolution (with revival of related issue #171.

**Comment by tdrozdovsky** [comment]
> 
> 
> > @tdrozdovsky Thank you for this PR. The suggested one is to resolve the overwritten attribute `err` in the given functions, am I correct? (`err` vs. `errn`) We are not sure how LGTM works in their analysis, so could you give us a little further information like : LGTM analysis result in your forked branch?
> 
> @tdrozdovsky Sorry that making you confused. LGTM does not provide any analysis on the forked version. So my suggestion is:
> 
>     1. Review and revise this PR here;
> 
>     2. (If we adopt this PR) let us see the updated LGTM analysis result;
> 
>     3. Decision as follows;
>        3-1. (If it works) Good;
>        3-2. (If it does not work) We might revoke this PR and find other resolution with revival of related issue #171.
> 
> 
> What do you think?

@MoonkiHong 
1. Yes you are right, I cannot check on the fork
2. I have no exact information that the errors will be fixed (but I'm sure it will work)
3. The LGTM once a day downloads the updated code from the master and updates the status (even if you run it manually, you will see the results later). Therefore, we have no other choice, only merge the code and wait a day (as I wrote earlier, this can be done only if we automate the verification process with each new PR).

In any case, I will keep this issue under control.
And we also need to separate (create new) issue from 171 issue by separating them. It is not possible to do them at the same time.
Thanks you.

**Review by MoonkiHong — APPROVED** [review]
LGTM.

**Comment by tdrozdovsky** [comment]
> 
> 
> LGTM.

Cool! This is a very quick answer!

**Comment by t25kim** [comment]
I'm not sure but I think this might be a LGTM's problem. Check this out with Peter's idea.

**Comment by tdrozdovsky** [comment]
> 
> 
> I'm not sure but I think this might be a LGTM's problem. Check this out with Peter's idea.

@t25kim you can check the analogous code and you will see that the local variable exists only for the time it is in the `case` block. The LGTM indirectly indicates this problem in the code.
``` 
package main

import "fmt"

func newValue(counter int) (int, error) {
	counter += 1
	errn := fmt.Errorf("New error value")
	return counter, errn
}

func main() {
	var err error
	err = fmt.Errorf("Old error value")
	a := 1
	switch a {
	case 1:
		b, err := newValue(a)
		fmt.Printf("a = %d; b = %d; error = %v\n", a, b, err)
	case 2:
		b, err := newValue(a)
		fmt.Printf("a = %d; b = %d; error = %v\n", a, b, err)
	case 3:
		b, err := newValue(a)
		fmt.Printf("a = %d; b = %d; error = %v\n", a, b, err)
	}
	fmt.Printf("a = %d; error = %v\n", a, err)
}
```
Result:
```
virtual-pc@virtualpc-VirtualBox:~/projects/gotest$ go run main.go 
a = 1; b = 2; error = New error value
a = 1; error = Old error value
```

**Comment by t25kim** [comment]
@tdrozdovsky  Thanks for the test code. but I think you should change the code to '=' instead of ':=' like below .

```
package main

import "fmt"

func newValue(counter int) (int, error) {
        counter += 1
        errn := fmt.Errorf("New error value")
        return counter, errn
}

func main() {
        var err error
        err = fmt.Errorf("Old error value")
        a := 1
        b := 2
        switch a {
        case 1:
                b, err = newValue(a)
                fmt.Printf("a = %d; b = %d; error = %v\n", a, b, err)
        case 2:
                b, err = newValue(a)
                fmt.Printf("a = %d; b = %d; error = %v\n", a, b, err)
        case 3:
                b, err = newValue(a)
                fmt.Printf("a = %d; b = %d; error = %v\n", a, b, err)
        }
        fmt.Printf("a = %d; error = %v\n", a, err)
}
```

The result is 
```
(venv) t25kim@t25kim:~/uni$ ./uni
a = 1; b = 2; error = New error value
a = 1; error = New error value
```

**Comment by t25kim** [comment]
@tdrozdovsky Sorry, you are right, I missed the code like "val, err := cast.ToBoolE(reading)"

**Review by t25kim — APPROVED** [review]
LGTM!

**Review by MoonkiHong — APPROVED** [review]
LGTM.

File: src/controller/discoverymgr/mnedc/client/client.go
Comment (t25kim) [inline src/controller/discoverymgr/mnedc/client/client.go:353]: I'm wondering this is right approach since this code is in the loop and other some [code](https://github.com/lf-edge/edge-home-orchestration-go/blob/7e9ed35d0cbdf46f9975c5bce5c2315201355eca/src/controller/discoverymgr/mnedc/client/client.go#L323) call **continue** instead of return.

**Review by t25kim — COMMENTED** [review]
Please see my comment and share your idea with us.

File: src/controller/discoverymgr/mnedc/client/client.go
Comment (tdrozdovsky) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: > 
> 
> I'm wondering this is right approach since this code is in the loop and other some [code](https://github.com/lf-edge/edge-home-orchestration-go/blob/7e9ed35d0cbdf46f9975c5bce5c2315201355eca/src/controller/discoverymgr/mnedc/client/client.go#L323) call **continue** instead of return.

Thanks for a very good question and really needs to be discussed.
My position is as follows: errors should be handled.

1. `err = tunIns.SetTUNIP (c.intf.Name (), c.virtualIP, c.netMask, true)`
does the following: assign an `c.virtualIP` address, `c.netMask` address for the `c.intf.Name ()`interface. 

General example for better understanding:
```
$ ifconfig eth0 172.16.25.125 netmask 255.255.255.22
```

2. `err = tunIns.SetTUNStatus(c.intf.Name(), true, true)`
performs the activation of the `c.intf.Name()` interface.

General example for better understanding:
```
$ ip link set dev eth0 up mtu 1400 qlen 300
```

**Can we activate the interface (command #2) if we have an error setting addresses (command #1)?**
If I misunderstood something or you have other thoughts, please write to me

Comment (tdrozdovsky) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: @MoonkiHong @t25kim Of course, if we are considering working in a loop, then we need to think about replacing the following `return` 
```
		intf, err := tunIns.CreateTUN()
		if err != nil {
			log.Println(logPrefix, "TUN error:", err.Error())
			return
		}

		c.intf = intf
		err = c.ParseVirtualIP(string(readBuf[0:n])) //unique TUN ip sent by server
		if err != nil {
			log.Println(logPrefix, "Parse Virtual IP error:", err.Error())
			return
		}

		err = tunIns.SetTUNIP(c.intf.Name(), c.virtualIP, c.netMask, true)
		if err != nil {
			log.Println(logPrefix, "TUN error:", err.Error())
			return
		}

		err = tunIns.SetTUNStatus(c.intf.Name(), true, true)
		if err != nil {
			log.Println(logPrefix, "TUN error:", err.Error())
			return
		}
```
with a continuation with the output of the corresponding error message
example:
```
if err != nil {
			log.Println(logPrefix, <ERROR>", retrying")
			time.Sleep(retryDelay)
			conn.Close()
			continue
		}
```

Comment (MoonkiHong) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: @tdrozdovsky Thank you for your kind analysis. We have also been waiting for the original `MNEDC` proposal developer from @suresh-lc and @sun-sharma. I personally think that we might be able to handle the logic issue in a separated PR linking with the registered (dedicated) issue #196. @tdrozdovsky @suresh-lc @t25kim What do you think?

Comment (tdrozdovsky) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: @MoonkiHong yes, I agree with you to allocate to another pool request. Give me your consent and I will remove this code from this pull request and create another one to fix this code.

P.S. It would be nice for us to close alerts from the LGTM faster!

Comment (MoonkiHong) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: @tdrozdovsky This PR is to resolve the reported LGTM issues. Let us merge this PR once all the reviewers approve this. Then let us create a separate PR to handle the logic issue as you requested. As we suggested, it would be better for us to have every PR with a single feature/enhancement from https://github.com/lf-edge/edge-home-orchestration-go/blob/master/.github/CONTRIBUTING.md (Section 5).

Comment (tdrozdovsky) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: @MoonkiHong Currently, we have 17 alerts. We will not be able to fix all with one pull request, since they are of different types. And I think that we will fix them step by step as there are more complex that will be considered for a long time, and also some that require some modification before fixing them. 

And I agree with your comment: "We have also been waiting for the original MNEDC proposal developer from @suresh-lc and @sun-sharma"

Comment (MoonkiHong) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: > @MoonkiHong Currently, we have 17 alerts. We will not be able to fix all with one pull request, since they are of different types. And I think that we will fix them step by step as there are more complex that will be considered for a long time, and also some that require some modification before fixing them.
> 
> And I agree with your comment: "We have also been waiting for the original MNEDC proposal developer from @suresh-lc and @sun-sharma"

@tdrozdovsky Seems that you agree with my proposal. Let us go ahead step-by-step. ^__^

Comment (t25kim) [reply src/controller/discoverymgr/mnedc/client/client.go:353]: @tdrozdovsky Thank you for sharing your idea with us. I know your code has no problem at the moment. I left the comment just for the future work.

**Review by t25kim — APPROVED** [review]
LGTM!

**Comment by MoonkiHong** [comment]
@suresh-lc @mgjeong Please take a look, Dear. ^___^

**Comment by mgjeong** [comment]
I don't have any experience related with LGTM analysis. So, I think we need the consent among all contributors about the principle what analysis-tools are used for our repository.

**Comment by MoonkiHong** [comment]
> I don't have any experience related with LGTM analysis. So, I think we need the consent among all contributors about the principle what analysis-tools are used for our repository.

@mgjeong Thank you for your prompt feedback. With my humble opinion, LGTM is getting a strong support from major open source contributing companies including Google and Microsoft, and it is recognized as one of the most reputation code-analysis services including security vulnerability in conjunction with CWE maintained by MITRE. Aside from LGTM itself, the common baseline artifact to execute the analysis is absed on CodeQL, and this is commonly used from similar other analysis tools provided currently from GitHub Action (as @t25kim). 

In summary, we see that the LGTM provides a trustworthy analysis for the security. So we should stop relying on this, what about creating the concerning separate issue and establish the principle for our project first, and keep the current usage of the LGTM before its completion? ^^ (Just my 2 cents)

**Review by suresh-lc — APPROVED** [review]
LGTM

**Comment by MoonkiHong** [comment]
> > I don't have any experience related with LGTM analysis. So, I think we need the consent among all contributors about the principle what analysis-tools are used for our repository.
> 
> @mgjeong Thank you for your prompt feedback. With my humble opinion, LGTM is getting a strong support from major open source contributing companies including Google and Microsoft, and it is recognized as one of the most reputation code-analysis services including security vulnerability in conjunction with CWE maintained by MITRE. Aside from LGTM itself, the common baseline artifact to execute the analysis is based on CodeQL, and this is commonly used from similar other analysis tools provided currently from GitHub Action (as @t25kim).
> 
> In summary, we see that the LGTM provides a trustworthy analysis for the security. So if you think that we should stop relying on this: what about creating the concerning separate issue and establish the principle for our project first, and keep the current usage of the LGTM before its completion? ^^ (Just my 2 cents)

I have created the regarding new issue (https://github.com/lf-edge/edge-home-orchestration-go/issues/199) to consider further for the principle of adopting the analysis tools. Thank you @mgjeong for your keen opinion. Once we complete establishing the principle, then we might be able to make a decision to keep or to change our current tools to alternatives. So before everything is completed, I would like to merge this PR and keep the current analysis tool as a tentative. Let me follow up this action from the separated GitHub issue (https://github.com/lf-edge/edge-home-orchestration-go/issues/199).

**Pull request merged**
